src/webgpu/api/validation/copyBetweenLinearDataAndTexture.spec.ts

```diff
@@ -0,0 +1,745 @@
+        dataSize: 1,
+        method: method,
+      });
+    }, !success);
+  });
+
+/*** Data-related validation: ***/
+
+g.test('texel_block_alignment_on_offset')
+  .params(
+    params()
+      .combine(poptions('method', [TestMethod.WriteTexture, TestMethod.CopyBufferToTexture, TestMethod.CopyTextureToBuffer]))
+      .combine([
+        { mul: 0, div: 1 }, // test 0
+        { mul: 1, div: 2 }, // test bytesPerBlock / 2
+        { mul: 1, div: 1 }, // test bytesPerBlock
+        { mul: 2, div: 1 }, // test 2 * bytesPerBlock
+        { mul: 3, div: 2 }, // test 3 * bytesPerBlock / 2
+        { mul: 3, div: 1 }, // test 3 * bytesPerBlock
+      ])
+      .combine(poptions('format', kTextureFormats))
+      .unless(({ format }) => !kTextureFormatInfo[format].copyable)
+  )
+  .fn(async t => {
+    const { mul, div, format, method } = t.params;
+
+    const offset = Math.floor(kTextureFormatInfo[format].bytesPerBlock * mul / div);
+
+    const texture = t.device.createTexture({
+      size: { width: 12, height: 12, depth: 12 },
+      mipLevelCount: 1,
+      format: format,
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST,
+    });
+
+    let success = offset % kTextureFormatInfo[format].bytesPerBlock === 0;
+
+    t.expectValidationError(() => {
+      t.testRun({
+        texture: texture,
+        origin: { x: 0, y: 0, z: 0 },
+        size: { width: 0, height: 0, depth: 0 },
+        offset: offset,
+        bytesPerRow: 1024,
+        rowsPerImage: 16,
+        dataSize: offset + 1,
+        method: method,
+      });
+    }, !success);
+  });
+
+g.test('texel_block_alignemnt_on_rows_per_image')
+  .params(
+    params()
+      .combine(poptions('method', [TestMethod.WriteTexture, TestMethod.CopyBufferToTexture, TestMethod.CopyTextureToBuffer]))
+      .combine(poptions('rowsPerImage', kTestValuesForDivisibilityBy4))
+      .combine(poptions('format', kTextureFormats))
+      .unless(({ format }) => !kTextureFormatInfo[format].copyable)
+  )
+  .fn(async t => {
+    const { rowsPerImage, format, method } = t.params;
+
+    const texture = t.device.createTexture({
+      size: { width: 12, height: 12, depth: 12 },
+      mipLevelCount: 1,
+      format: format,
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST,
+    });
+
+    let success = rowsPerImage % kTextureFormatInfo[format].blockHeight === 0;
+
+    t.expectValidationError(() => {
+      t.testRun({
+        texture: texture,
+        origin: { x: 0, y: 0, z: 0 },
+        size: { width: 0, height: 0, depth: 0 },
+        offset: 0,
+        bytesPerRow: 1024,
+        rowsPerImage: rowsPerImage,
+        dataSize: 1,
+        method: method,
+      });
+    }, !success);
+  });
+
+g.test('bound_on_rows_per_image')
+  .params(
+    params()
+      .combine(poptions('method', [TestMethod.WriteTexture, TestMethod.CopyBufferToTexture, TestMethod.CopyTextureToBuffer]))
+      .combine(poptions('rowsPerImageInBlock', [0, 1, 2]))
```

Comment:
FYI we are just changing rowsPerImage to be measured in blocks: https://github.com/gpuweb/gpuweb/pull/958

Reply:
Ugh, when will that have to be reflected in the CTS?

Reply:
Anytime is fine, we definitely don't always stay in lock step with the spec or the implementations. We can do it now or leave it as TODO.

In dawn and chrome I think we are going to do this as a breaking change (instead of a deprecation) as doing a deprecation here is hard.